include/private/rdycoreimpl.h

```diff
@@ -28,49 +129,126 @@ struct _p_RDy {
   // TODO: The fields below are subject to change as we find our way!
   //------------------------------------------------------------------
 
-  /// MPI communicator used for the simulation
+  // MPI communicator used for the simulation
   MPI_Comm comm;
-  /// MPI rank of local process
+  // MPI rank of local process
   PetscInt rank;
-  /// Number of processes in the communicator
-  PetscInt comm_size;
-  /// filename storing input data for the simulation
-  char filename[PETSC_MAX_PATH_LEN];
-  /// PETSc grid
+  // Number of processes in the communicator
+  PetscInt nproc;
+  // file storing input data for the simulation
+  char config_file[PETSC_MAX_PATH_LEN];
+
+  //---------
+  // Physics
+  //---------
+
+  // flow and related parameterization(s)
+  RDyFlowMode    flow_mode;
+  RDyBedFriction bed_friction;
+  PetscReal      bed_friction_coef;
+
+  // sediment
+  PetscBool sediment;
+
+  // salinity
+  PetscBool salinity;
+
+
+  //----------
+  // Numerics
+  //----------
+  RDySpatial spatial;
+  RDyTemporal temporal;
+  RDyRiemann riemann;
+
+  //------------------------
+  // Spatial discretization
+  //------------------------
+
+  // mesh file
+  char mesh_file[PETSC_MAX_PATH_LEN];
+
+  // PETSc (DMPlex) grid
   DM dm;
-  /// Number of cells in the x direction
-  PetscInt Nx;
-  /// Number of cells in the y direction
-  PetscInt Ny;
-  /// grid spacing in the x direction
-  PetscReal dx;
-  /// grid spacing in the y direction
-  PetscReal dy;
-  /// domain extent in x
-  PetscReal Lx;
-  /// domain extent in y
-  PetscReal Ly;
-  /// water depth for the upstream of dam [m]
-  PetscReal hu;
-  /// water depth for the downstream of dam [m]
-  PetscReal hd;
-  /// water depth below which no horizontal flow occurs
-  PetscReal tiny_h;
-  /// total number of time steps
-  PetscInt Nt;
-  /// time step size
+
+  // mesh representing simulation domain
+  RDyMesh mesh;
+
+  // mesh regions
+  PetscInt  num_regions;
+  PetscInt  region_ids[1+MAX_REGION_ID];
+  RDyRegion regions[1+MAX_REGION_ID];
+
+  // mesh surfaces
+  PetscInt   num_surfaces;
+  PetscInt   surface_ids[1+MAX_SURFACE_ID];
+  RDySurface surfaces[1+MAX_SURFACE_ID];
+
+  // Table of named sets of flow, sediment, and salinity conditions. Used by
+  // initial/source/boundary conditions below.
+  PetscInt             num_flow_conditions;
+  RDyFlowCondition     flow_conditions[1+MAX_CONDITION_ID];
```

Comment:
Why not make this a pointer and allocate the appropriate size depending on the entries in the yaml file instead of using `MAX_CONDITION_ID`?

Reply:
Good question. I did this because we read in these conditions one at a time because of the way input is parsed. In order to allocate the right number of conditions, we'd have to read everything into an intermediate representation beforehand, which I decided not to do because it's more code. I think this will work well enough to get us started. If we teach PETSc's options database to parse this kind of YAML, it will serve as that intermediate representation and then we can do what you propose.

Reply:
sounds good.